docs/tutorials/adding-a-module-to-your-runtime.md

```diff
@@ -0,0 +1,306 @@
+        Block = Block,
+        NodeBlock = opaque::Block,
+        UncheckedExtrinsic = UncheckedExtrinsic
+    {
+        ...
+        Contract: contract::{Module, Call, Storage, Event<T>, Config<T>},
+    }
+);
+```
+
+### Adding Runtime Hooks
+
+Substrate provides the ability for modules to expose "hooks" where changes in the module can trigger functions in other modules. For example, you could create a module which executes some action every time a new account is created (when it first gains a balance over the [existential deposit](https://substrate.dev/docs/en/overview/glossary#existential-deposit)).
+
+In the case of the Contracts module, we actually want a hook when an account runs out of free balance. Because the Contract module instantiates contracts as "Accounts", it also needs to know when an account is destroyed so that it can clean up any storage that contract was using. You can find that logic in the Contract module source code:
+
+**Contract module `lib.rs`**
+```rust
+impl<T: Trait> OnFreeBalanceZero<T::AccountId> for Module<T> {
+	fn on_free_balance_zero(who: &T::AccountId) {
+		<CodeHashOf<T>>::remove(who);
+		<AccountInfoOf<T>>::get(who).map(|info| child::kill_storage(&info.trie_id));
+	}
+}
+```
+
+To enable this, we simply need to add `Contract` to the `OnFreeBalanceZero` hook provided by the Balances module:
+
+**`contract-chain/runtime/src/lib.rs`**
+```rust
+impl balances::Trait for Runtime {
+    /// What to do if an account's free balance gets zeroed.
+    type OnFreeBalanceZero = (Contract);
+    ...
+}
+```
+
+Now, when the Balances module detects that the free balance of an account has reached zero, it calls the `on_free_balance_zero` function of the contract module.
+
+## Genesis Configuration
+
+The last thing we need to do in order to get your node up and running is to establish a genesis configuration for the Contract module. This is controlled in `contract-chain/src/chain_spec.rs`. We need to modify this file to include the `ContractConfig` type:
+
+```rust
+use contract_chain_runtime::ContractConfig;
+```
+
+Then we need to add values to the configuration. As of Substrate v1.0, it looks like:
+
+```rust
+let mut contract_config = ContractConfig {
+    transaction_base_fee: 1,
+    transaction_byte_fee: 0,
+    transfer_fee: 0,
+    creation_fee: 0,
+    contract_fee: 21,
+    call_base_fee: 135,
+    create_base_fee: 175,
+    gas_price: 1,
+    max_depth: 1024,
+    block_gas_limit: 10_000_000,
+    current_schedule: Default::default(),
+};
+// IMPORTANT: this should only be enabled on development chains!
+contract_config.current_schedule.enable_println = true;
+```
+
+Note that you can tweak these numbers to your needs, but these are the values set at the genesis configuration of the main Substrate node. You should place that code block into the `testnet_genesis` function, and then in the `GenesisConfig` object, add:
+
+```rust
+GenesisConfig {
+    ...
+    contract: Some(contract_config),
+}
+```
```

Comment:
Defining `contract_config` variable and then using it is slightly different than how other modules are done in the node template. Learners may not realize that they've done basically the same thing as:

```rust
GenesisConfig {
        contract: Some(ContractConfig {
            transaction_base_fee: 1,
            transaction_byte_fee: 0,
            transfer_fee: 0,
            creation_fee: 0,
            contract_fee: 21,
            call_base_fee: 135,
            create_base_fee: 175,
            gas_price: 1,
            max_depth: 1024,
            block_gas_limit: 10_000_000,
            current_schedule: Default::default(),
        }),
        ...
}
```

I see that this technique is necessary for
```
// IMPORTANT: this should only be enabled on development chains!
contract_config.current_schedule.enable_println = true;
```

I can guess from the name that's something about allowing printing to the console from within contracts? Will I encounter that anywhere else? Maybe worth just mentioning that what we're doing here isn't really that different from how the other modules are doing it?

Reply:
I think we should keep it in since as I mentioned, it is necessary if you actually want to develop contracts on Substrate.

I agree though, a bit of a bummer not to follow the same pattern.